### PR TITLE
feat: add public input validation logic

### PR DESCRIPTION
**Pull request by chokobole**
Related: #17

### CODE REVIEW

**Review by Pangssu — APPROVED** [review]
lgtm

**Comment by chokobole** [comment]
> The test `TestFees` fails in `e2e/system_test.go` due to the changes in fee distribution. Let's fix the test in this PR and then merge it into `dev`.

Okay, then i'll look into this.

**Comment by 0xHansLee** [comment]
> The test TestFees fails in e2e/system_test.go due to the changes in fee distribution.

Sorry for confusing you. This PR is quite simple and the test failed is not due to this change.
Since we are fixing the E2E tests related to validator decentralization, why don't we merge this PR first and handle it in the E2E fixing PR?

**Review by 0xHansLee — APPROVED** [review]
> Sorry for confusing you. This PR is quite simple and the test failed is not due to this change.
Since we are fixing the E2E tests related to validator decentralization, why don't we merge this PR first and handle it in the E2E fixing PR?

It was resolved by rebasing onto dev. LGTM

File: specs/challenge.md [hunk 1]
```diff
@@ -214,10 +214,13 @@ verification by [ZK Verifier Contract](./zkevm-prover.md#the-zk-verifier-contrac
 1. Check whether `_srcOutputRootProof` is the preimage of the first output root of the segment.
 2. Check whether `_dstOutputRootProof` is the preimage of the next output root of the segment.
 3. Verify that the `nextBlockHash` in `_srcOutputRootProof` matches the `blockHash` in `_dstOutputRootProof`.
-4. Verify that the `blockHash` in `_srcOutputRootProof` matches the block hash derived from `_publicInput` and `_rlps`.
-5. Verify that the `transactionsRoot` in `_publicInput` matches the transaction root derived from the transaction
+4. Verify that the `stateRoot` in `_publicInput` matches the `stateRoot` in `_dstOutputRootProof`.
+5. Verify that the `blockHash` in `_srcOutputRootProof` matches the block hash derived from `_publicInput` and `_rlps`.
```
Comment (seolaoh) [inline specs/challenge.md:218]:
```suggestion
5. Verify that the `nextBlockHash` in `_srcOutputRootProof` matches the block hash derived from `_publicInput` and `_rlps`.
```

File: specs/challenge.md [hunk 2]
```diff
@@ -214,10 +214,13 @@ verification by [ZK Verifier Contract](./zkevm-prover.md#the-zk-verifier-contrac
...
+6. Verify that the `transactionsRoot` in `_publicInput` matches the transaction root derived from the transaction
    hashes.
-6. If the transaction hash in `_publicInput` is less than `MAX_TXS`, fill it with `DUMMY_HASH`.
+7. Verify that the `withdrawalStorageRoot` in `_dstOutputRootProof` is contained in `stateRoot` in
+  `_dstOutputRootProof`.
+8. If the transaction hash in `_publicInput` is less than `MAX_TXS`, fill it with `DUMMY_HASH`.
```
Comment (seolaoh) [inline specs/challenge.md:223]:
```suggestion
8. If the length of transaction hashes in `_publicInput` is less than `MAX_TXS`, fill it with `DUMMY_HASH`.
```

**Review by seolaoh — APPROVED** [review]
LGTM

**Pull request merged**